psl/psl-core/src/validate/datasource_loader.rs

```diff
@@ -185,47 +189,77 @@ fn lift_datasource(
     })
 }
 
+/// Returns the relation mode for the datasource, validating against invalid relation mode settings and
+/// the deprecated `referentialIntegrity` attribute.
 fn get_relation_mode(
     args: &mut HashMap<&str, (Span, &ast::Expression)>,
     source: &SourceConfig,
     diagnostics: &mut Diagnostics,
     connector: &'static dyn crate::datamodel_connector::Connector,
 ) -> Option<RelationMode> {
-    match (args.remove("relationMode"), args.remove("referentialIntegrity")) {
-        (None, None) => None,
-        (Some(_), Some((span, _))) => {
-            diagnostics.push_error(DatamodelError::new_referential_integrity_and_relation_mode_cooccur_error(span));
-            None
-        }
-        (Some((_span, rm)), None) | (None, Some((_span, rm))) => {
-            let integrity = match coerce::string(rm, diagnostics)? {
```

Comment:
Moved the code down here to the new `get_and_validate_relation_mode_value` and `validate_allowed_relation_mode_settings` closures, as they're used both when only the `relationMode` attribute is used or when only the `referentialIntegrity` attribute is used

Reply:
I think we are mixing concerns and making the code more complicated here. Would something much simpler like `if args.contains("referentialIntegrity") { ... push the warning here ... }` at the beginning of the function, before the `match` works. It would emit two warnings in case you defined both, but that's fine.

Reply:
Ok, done